design/OrthogonalPersistence.md

```diff
@@ -0,0 +1,126 @@
+# Orthogonal Persistence (Stable Heap)
+
+This realizes the vision of keeping the canister main memory persistent even across upgrades and thus allows scalable upgrades.
+Canister upgrades do no longer involve serialization and deserialization to and from secondary stable memory.
+
+## Purpose
+* **Instantenous upgrades**: New program versions simply resume on the existing main memory and have access to the memory-compatible data.
+* **Scalable upgrades**: The upgrade mechanism scales with larger heaps and in contrast to serialization, does not hit IC instruction limits.
+
+## Broader Vision
+In the longer term, this approach aims to enable **true orthogonal persistence** that is simple, flexible, efficient, and scalable.
+While this version implements the runtime support for 32-bit memory, this could be leveraged to 64-bit persistent main memory in future.
+As a result, the use of secondary storage (explicit stable memory, dedicated stable data structures, DB-like storage abstractions) will no longer be necessary: 
+Motoko developers could directly work on their normal object-oriented program structures that are automatically persisted and retained across program version changes.
+With 64-bit main memory, large-scaled orthogonal persistence would be enabled, supported by the incremental GC that is designed to also scale in 64-bit.
+
+## Design
+The stable heap is based on the following main properties:
+* Extension of the IC to retain main memory on upgrades.
+* A long-term memory layout that is invariant to new compiled program versions.
+* A fast memory compatibility check performed on each canister upgrade.
+* Incremental garbage collection using a partitioned heap.
+
+### IC Extension
+As a prerequisite for the stable heap support, the IC runtime support has to be extended in order not to erase the main memory on upgrades.
+This is realized in a specific IC PR (https://github.com/luc-blaeser/ic/tree/luc/stable-heap-on-release) that retains the main memory even on upgrades, similar to normal canister message execution. 
+
+### Memory Layout
+In a co-design between the compiler and the runtime system, the main memory is arranged in the following structure, invariant of the compiled program version:
+* Lower 4MB: Rust call stack.
+* Space between 4MB and 4.5MB: Limited reserved space Wasm data segments, only used for the Motoko runtime system.
```

Comment:
Again,  I wonder if this is too limited. What happens if we add proper support for unicode to the rts? I think that might increase the data segments by up to 1MB (though hard to tell without doing the experiment). 
If we get stuck, does that mean we can get unstuck by doing a graph copy upgrade?

Reply:
Yes, it is hard to tell what a good limit is. We would have two options: 
* Make Rust use a passive data segment, e.g. importing such functionality in Rust. 
* Or doing a memory layout change with graph copy.
I believe the first approach would the more elegant one, especially if it is a bulk data segment.